examples/imports_function_env_global.rs

```diff
@@ -0,0 +1,156 @@
+//!
+//!   1. There will be a `get_counter` function that will return an i32 of
+//!      the current global counter, The function will also increment a global value
+//!   2. There will be an `add_to_counter` function will add the passed
+//!      i32 value to the counter, and return an i32 of the current
+//!      global counter.
+//!
+//! You can run the example directly by executing in Wasmer root:
+//!
+//! ```shell
+//! cargo run --example imported-function-env-global --release --features "cranelift"
+//! ```
+//!
+//! Ready?
+
+use std::sync::{Arc, Mutex};
+use wasmer::{
+    imports, wat2wasm, Function, FunctionEnv, FunctionEnvMut, Global, Instance, Module, Store,
+    TypedFunction, Value,
+};
+use wasmer_compiler_cranelift::Cranelift;
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    // Let's declare the Wasm module with the text representation.
+    let wasm_bytes = wat2wasm(
+        br#"
+(module
+  (global $g_counter (import "env" "g_counter") (mut i32))
+  (func $get_counter (import "env" "get_counter") (result i32))
+  (func $add_to_counter (import "env" "add_to_counter") (param i32) (result i32))
+
+  (type $increment_t (func (param i32) (result i32)))
+  (func $increment_f (type $increment_t) (param $x i32) (result i32)
+    (block
+      (loop
+        (call $add_to_counter (i32.const 1))
+        (set_local $x (i32.sub (get_local $x) (i32.const 1)))
+        (br_if 1 (i32.eq (get_local $x) (i32.const 0)))
+        (br 0)))
+    call $get_counter)
+  (export "increment_counter_loop" (func $increment_f)))
+"#,
+    )?;
+
+    // Create a Store.
+    // Note that we don't need to specify the engine/compiler if we want to use
+    // the default provided by Wasmer.
+    // You can use `Store::default()` for that.
+    let mut store = Store::new(Cranelift::default());
+
+    println!("Compiling module...");
+    // Let's compile the Wasm module.
+    let module = Module::new(&store, wasm_bytes)?;
+
+    // Create the global
+    let g_counter = Global::new_mut(&mut store, Value::I32(5));
+
+    // We create some shared data here, `Arc` is required because we may
+    // move our WebAssembly instance to another thread to run it. Mutex
+    // lets us get shared mutabilty which is fine because we know we won't
+    // run host calls concurrently.  If concurrency is a possibilty, we'd have
+    // to use a `Mutex`.
+    let shared_counter: Arc<Mutex<i32>> = Arc::new(Mutex::new(0));
+
+    // Once we have our counter we'll wrap it inside en `Env` which we'll pass
+    // to our imported functionsvia the FunctionEnv.
+    //
+    // This struct may have been anything. The only constraint is it must be
+    // possible to know the size of the `Env` at compile time (i.e it has to
+    // implement the `Sized` trait).
+    // The Env is then accessed using `data()` or `data_mut()` method.
+    #[derive(Clone)]
+    struct Env {
+        counter: Arc<Mutex<i32>>,
```

Comment:
Why this do need to be a mutex?

Reply:
Just read the comment

Reply:
But I'm not sure if I agree with it, we can discuss in a call

Reply:
It's based on the `import_function_env` example, it's really mostly copy/paste. But we can have a call about the examples for sure.